docs/design.md

```diff
@@ -0,0 +1,371 @@
+  * `[Lock/Unlock]Shard()`: for exclusive access.
+  * `[Incr/Decr]Shard()`: refcounting on shards.
+
+### Mounts
+
+Shards can be located anywhere, and can come and go dynamically e.g. Filecoin
+deals expire, removable media is attached/detached, or the IPFS user purges
+content.
+
+It is possible to mount shards with CARs accessible through the local
+filesystem, detachable mounts, NFS mounts, distributed filesystems like
+Ceph/GlusterFS, HTTP, FTP, etc.
+
+This versatility is provided by an abstraction called `mount.Mount`, a pluggable
+component which encapsulates operations/logic to:
+
+1. `Fetch() (io.ReadCloser, error)` Load a CAR from its origin.
+2. `Info() mount.Info` Provides info about the mount, e.g. whether it's local or
+   remote. This is used to determine whether the fetched CAR needs to be copied
+   to a scrap area.
+3. `Stat() (mount.Stat, error)` Equivalent to a filesystem stat, provides
+   information about the target of the mount: whether it exists, size, etc.
+
+When instantiating `Mount` implementations, one can provide credentials, access
+tokens, or other parameters through the implementation constructors. This is
+necessary if access to the CAR is permissioned/authenticated.
+
+**Local filesystem Mount**
+
+A local filesystem `Mount` implementation loads the CAR directly from the
+filesystem file. It is of `local` type and therefore requires no usage of scrap
+area.
+
+*This `Mount` is provided out-of-box by the DAG store.*
+
+**Lotus Mount implementation**
+
+A Lotus `Mount` implementation would be instantiated with a sector ID and a
+bytes range within the sealed sector file (i.e. the deal segment).
+
+Loading the CAR consists of calling the worker HTTP API to fetch the unsealed
+piece. Because the mount is of `remote` type, the DAG store will need to store
+it in a local scrap area. Currently, this may lead to actual unsealing on the
+Lotus worker cluster through the current PoRep (slow) if the piece is sealed.
+
+With a future PoRep (cheap, snappy) PoRep, sealing can be performed _just_ for
+the blocks that are effectively accessed, potentially during IPLD block access
+time. A transformation function may be provided in the future as a `RegisterOpt`
+that conducts the unsealing.
+
+A prerequisite to enable unsealing-on-demand possible is PoRep and index
+symmetry, i.e. the byte positions of blocks in the unsealed CAR must be
+identical to those in the unsealed CAR.
+
+*This `Mount` is provided by Lotus, as it's implementation specific.*
+
+### Shard representation and persistence
+
+The shard catalogue needs to survive restarts. Thus, it needs to be persistent.
+Options to explore here include LMDB, BoltDB, or others. Here's what the
+persisted shard entry could look like:
+
+```go
+type PersistedShard struct {
+   Key   []byte
+   // Mount is a URL representation of the Mount, e.g.
+   //   file://path/to/file?opt=1&opt=2
+   //   lotus://sectorID?offset=1&length=2
+   Mount string
```

Comment:
1. To make it easy to persist Lotus mount implementations,  we could build out an endpoint in Markets that takes care of the magic elaborated [above](https://github.com/filecoin-project/dagstore/pull/2/files#r655842956). We can then persist this endpoint.

2. I think a better idea is to leave the details of Mount persistence to callers as mentioned in [this comment](https://github.com/filecoin-project/dagstore/pull/2/files#r655875242) for future extensibility.

Reply:
We don't need to persist behaviours, we need to persist state and have a mechanism to bind state to their implementations. Luckily, `Mounts` are conceived as stateless entities, so we only really need to persist the _configuration_ of a mount, which is done via URLs.

I've fleshed out more about this in the design doc now.